packages/eslint-plugin/src/rules/dependency-checks.ts

```diff
@@ -0,0 +1,379 @@
+    ignoredDependencies?: string[];
+  }
+];
+
+export type MessageIds =
+  | 'noPackageJson'
+  | 'missingDependency'
+  | 'obsoleteDependency'
+  | 'versionMismatch'
+  | 'missingDependencySection';
+
+export const RULE_NAME = 'dependency-checks';
+
+export default createESLintRule<Options, MessageIds>({
+  name: RULE_NAME,
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: `Checks dependencies in project's package.json for version mismatches`,
+      recommended: 'error',
+    },
+    fixable: 'code',
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          buildTargets: [{ type: 'string' }],
+          ignoreDependencies: [{ type: 'string' }],
+          checkMissingDependencies: { type: 'boolean' },
+          checkObsoleteDependencies: { type: 'boolean' },
+          checkVersionMismatches: { type: 'boolean' },
+          checkMissingPackageJson: { type: 'boolean' },
+        },
+        additionalProperties: false,
+      },
+    ],
+    messages: {
+      noPackageJson: `The "package.json" is required for buildable libraries, but does not exist in "{{projectName}}".`,
+      missingDependency: `The "{{packageName}}" is missing from the "package.json".`,
+      obsoleteDependency: `The "{{packageName}}" is found in the "package.json" but is not used.`,
+      versionMismatch: `The "{{packageName}}" is found in the "package.json" but it's range is not matching the installed version. Installed version: {{version}}.`,
+      missingDependencySection: `The "dependencies" section is missing from the "package.json".`,
+    },
+  },
+  defaultOptions: [
+    {
+      buildTargets: ['build'],
+      checkMissingDependencies: true,
+      checkObsoleteDependencies: true,
+      checkVersionMismatches: true,
+      checkMissingPackageJson: true,
+      ignoredDependencies: [],
+    },
+  ],
+  create(
+    context,
+    [
+      {
+        buildTargets,
+        ignoredDependencies,
+        checkMissingDependencies,
+        checkObsoleteDependencies,
+        checkVersionMismatches,
+        checkMissingPackageJson,
+      },
+    ]
+  ) {
+    if (!(context.parserServices as any).isJSON) {
+      return {};
+    }
+    const fileName = normalizePath(context.getFilename());
+    // support only package.json and project.json
+    if (
+      !fileName.endsWith('/package.json') &&
+      !fileName.endsWith('/project.json')
+    ) {
+      return {};
+    }
+
+    const projectPath = normalizePath(
+      (global as any).projectPath || workspaceRoot
+    );
+    const sourceFilePath = getSourceFilePath(fileName, projectPath);
+    const { projectGraph, projectRootMappings, projectFileMap } =
+      readProjectGraph(RULE_NAME);
+
+    if (!projectGraph) {
+      return {};
+    }
+
+    const sourceProject = findProject(
+      projectGraph,
+      projectRootMappings,
+      sourceFilePath
+    );
+
+    // check if source project exists, it's library and has a build target
```

Comment:
This comment isn't accurate as is - the check doesn't look for a build target, that's later. We should update it, or remove it since its not really providing extra clarity.

```suggestion
    // check if source project exists and its a library
```